feat: Introduce graphed, dependency based downloads

kristofbolyai opened the pull request:
`DownloadManager` has the documentation I'd put in the PR description.

TLDR; The goal is to eliminate download dependency race once and for all, all while taking benefit of having a manager behind all downloads, which include proper failure handling / retries, maximum parallel download settings, a clear and easily calculated download progress percentage and a list of download failures which can be shown to the user.

kristofbolyai committed:
wip: feat: Introduce graphed, dependency based downloads

`DownloadManager` has the documentation I'd put in the PR description.

TLDR; The goal is to eliminate download dependency race once and for all, all while taking benefit of having a manager behind all downloads, which include proper failure handling / retries, maximum parallel download settings, a clear and easily calculated download progress percentage and a list of download failures which can be shown to the user.

magicus commented on common/src/main/java/com/wynntils/core/net/DownloadManager.java:
```diff
@@ -0,0 +1,70 @@
+ * Manages downloading files from the internet. This manager acts as a middle layer between the caller components
+ * and {@link NetManager}. It's main purpose is to provide the components a well regulated and managed way
+ * of downloading dependency files.
+ * <p><b>This manager should not be used for spontainous downloads. Use one {@link NetManager}'s download methods
```
typo

magicus commented:
Oh, very nice!!!

I think you have the completely right idea here. The design sketched here looks sound; consider it approved. :)

But sometimes I wonder why we have to do so much ground-work ourselves. Some of the hardest part of the mod to get right have had nothing to do with Wynntils or Minecraft modding at all, like config/storage, downloading and event buses. *sigh* There is an old saying: "Any sufficiently large software project will end up re-implementing emacs", or something like that. I feel that is kind of what we are doing.

kristofbolyai committed:
feat: Introduce the API for registering data dependencies, move everything over to it

# Conflicts:
#	common/src/main/java/com/wynntils/models/gear/GearInfoRegistry.java
#	common/src/main/java/com/wynntils/models/ingredients/IngredientInfoRegistry.java
#	common/src/main/java/com/wynntils/models/rewards/CharmInfoRegistry.java
#	common/src/main/java/com/wynntils/models/rewards/RewardsModel.java
#	common/src/main/java/com/wynntils/models/rewards/TomeInfoRegistry.java

kristofbolyai commented:
Now you can check how the code API would look, without an actual backing implementation. I wanted to break away from how `NetManger#download` is always callable, so I introduced a proper time to register these downloads. I think they turned out great, and removed most of the previous mess we've had dealing with dependencies.

magicus commented on common/src/main/java/com/wynntils/models/ingredients/IngredientModel.java:
```diff
@@ -17,40 +16,23 @@
+
     private final IngredientInfoRegistry ingredientInfoRegistry = new IngredientInfoRegistry();
 
     public IngredientModel(WynnItemModel wynnItem) {
```
I'm guessing the dependency in wynnItem can be removed now? My reading of this code was that it was added either when we loaded the ingredient DB in the constructor, or as a safe guard against a race condition when it was made to be asynchronously loaded.

kristofbolyai replied:
That's likely correct. Or maybe when our "dependency" requirement was more strict. I'll remove it and check other cases as well.

magicus commented on common/src/main/java/com/wynntils/services/map/PoiService.java:
```diff
@@ -75,87 +74,21 @@ public class PoiService extends Service {
+        registry.registerDownload(UrlId.DATA_STATIC_PLACES).handleReader(this::handlePlaces);
+        registry.registerDownload(UrlId.DATA_STATIC_SERVICES).handleReader(this::loadServices);
+        registry.registerDownload(UrlId.DATA_STATIC_COMBAT_LOCATIONS).handleReader(this::loadCombat);
+        registry.registerDownload(UrlId.DATA_STATIC_CAVE_INFO).handleReader(this::handleCaves);
```
The naming is a bit inconsistent. Maybe rename all to "handle*"?

kristofbolyai replied:
Indeed. I think these were the only ones I missed though, I wanted all named handleX.

magicus commented:
This looks really, really great.

magicus commented:
We have some way to trigger a full re-download from a command. I recall not being entirely happy on how that was implemented. Maybe that is something that can be better integrated into the DownloadManager and made nicer as well?

kristofbolyai commented:
> We have some way to trigger a full re-download from a command. I recall not being entirely happy on how that was implemented. Maybe that is something that can be better integrated into the DownloadManager and made nicer as well?

Yes. It's wynntils reload / wynntils clearcaches that calls Model#reload or smth like that. It's definitely planned for renovation either in this PR, or as a follow up. I want all these download changes released as a minor version bump (likely 2.2.0, as 2.1.0 will be the Wynn V3.3 API update this weekend), so I'll retarget this PR to `feat/data-loading`.

kristofbolyai committed:
feat: Add the ability to download dependencies using the manager

kristofbolyai committed:
fix: fix accidentally left out renames and remove unused reloadData calls

kristofbolyai committed:
feat: make /wynntils reloadcaches work with DownloadManager

kristofbolyai commented:
@magicus This should be reviewable now. Now, the mod uses the new manager, but new features, automatic retry on error is not implemented. In its current form, it's a replacement for the old messy dependencies, except dependency management actually works. Reloading also works as intended. We can already calculate some statistics, like error rate, however it's only in logs at the moment. I plan on adding each new feature as a separate PR, while also fixing some older issues related to data loading, hence the new branch this PR targets.

After a lot of iterations on my end, I've decided to keep it simple. I've went through a lot of iterations with CompletableFuture, implementing custom NetResult types. Currently, I am happy with the code's current form, although that may change in the upcoming days, which is why I am keeping these batch of changes out of main for now.

kristofbolyai committed:
fix: fix typo, remove unused dependencies in models

kristofbolyai committed:
chore: revert NetResult changes

kristofbolyai committed:
fix: fix merge

# Conflicts:
#	common/src/main/java/com/wynntils/models/gear/GearInfoRegistry.java

kristofbolyai committed:
fix: Make DownloadDependencyGraph#nextDownload synchronized

kristofbolyai committed:
fix: tostring for dependencies, only log error in DownloadManager

magicus commented:
I'll try to look at this as soon as I can.

magicus commented on common/src/main/java/com/wynntils/core/net/DownloadDependencyGraph.java:
```diff
@@ -0,0 +1,286 @@
+        // 1. Check for duplicate downloads
+        checkDuplicateDownloads(downloads);
+
+        // 2. Check for missign dependencies
```
```suggestion
        // 2. Check for missing dependencies
```

magicus committed:
fix typo

magicus commented on common/src/main/java/com/wynntils/core/net/DownloadDependencyGraph.java:
```diff
@@ -0,0 +1,286 @@
+                    + " downloaded the same data twice: " + duplicatedInSameComponent.urlId());
+        }
+
+        // Downloading the same resouce in different core components is not a direct confirmation of a schematic issue,
```
```suggestion
        // Downloading the same resource in different core components is not a direct confirmation of a schematic issue,
```

magicus committed:
fix typo

magicus commented on common/src/main/java/com/wynntils/core/net/DownloadManager.java:
```diff
@@ -0,0 +1,220 @@
+        throw new IllegalStateException("Queued download has no handler set: " + queuedDownload);
+    }
+
+    private void replaceDownload(QueuedDownload toBeReplaced) {
```
The name got me thinking that you first downloaded to a temporary file name, and then moved the file in place one the download was successful. (If we don't do that elsewhere in the NetManager, it is probably something we should do.)

But after scratching my head a bit, I realize this is the code responsible for repopulating the queue. I'd suggest renaming the method. Maybe `queueNextDownload()` instead?

magicus commented on common/src/main/java/com/wynntils/services/map/PoiService.java:
```diff
@@ -75,87 +74,21 @@ public class PoiService extends Service {
-    @SubscribeEvent
-    public void onDataLoaded(NetResultProcessedEvent.ForUrlId event) {
-        if (!WynntilsMod.isDevelopmentEnvironment()) return;
-        // Serialize loading of POIs when on dev env
```
So this was just about triggering a race, which will be resolved by itself in the new system? Asking this as I see that you have added no dependency constraints for the DownloadManager. 

Or, to phrase it differently: are you sure that there are no dependencies between these downloads that you have missed to express?

kristofbolyai replied:
We really have two options here. The serialized loading for POIs originally happened because of a multi-thread class loading race between threads in the fork join pool. As far as I know it's been fixed in Arch-Transformer / EventBus (or Bus, as they call it now..). I can either add the dependency back, or what I would rather do, remove it, as it's really likely to be fixed, and even if it is not, it's a race that only happens in development environment, meaning users won't have issues with it.

magicus replied:
Right, it was some class loading stuff. Now I recall. Yeah, remove it, and hope it's been fixed. Otherwise we'll deal with it.

magicus reviewed (approved):
Overall it looks very good! I left a few comments/questions.

kristofbolyai committed:
refactor: rename method

kristofbolyai committed:
fix: Sync download object to prevent queued downloads finishing too early
